temporalio/client.py

```diff
@@ -5298,50 +5924,117 @@ async def start_workflow_update(
+                start_input
+            )
+        )
+        update_req = await self._build_update_workflow_execution_request(
+            update_input, workflow_id=start_input.id
+        )
+        multiop_req = temporalio.api.workflowservice.v1.ExecuteMultiOperationRequest(
+            namespace=self._client.namespace,
+            operations=[
+                temporalio.api.workflowservice.v1.ExecuteMultiOperationRequest.Operation(
+                    start_workflow=start_req
+                ),
+                temporalio.api.workflowservice.v1.ExecuteMultiOperationRequest.Operation(
+                    update_workflow=update_req
+                ),
+            ],
+        )
+
+        # Repeatedly try to invoke ExecuteMultiOperation until the update
+        # reaches user-provided wait stage or is at least ACCEPTED (as of the
+        # time of this writing, the user cannot specify sooner than ACCEPTED)
+        seen_start = False
         while True:
             try:
-                resp = await self._client.workflow_service.update_workflow_execution(
-                    req,
-                    retry=True,
-                    metadata=input.rpc_metadata,
-                    timeout=input.rpc_timeout,
+                multiop_response = (
+                    await self._client.workflow_service.execute_multi_operation(
+                        multiop_req
+                    )
                 )
+            except asyncio.CancelledError as err:
+                raise WorkflowUpdateRPCTimeoutOrCancelledError() from err
             except RPCError as err:
-                if (
-                    err.status == RPCStatusCode.DEADLINE_EXCEEDED
-                    or err.status == RPCStatusCode.CANCELLED
-                ):
+                if err.status in [
+                    RPCStatusCode.DEADLINE_EXCEEDED,
+                    RPCStatusCode.CANCELLED,
+                ]:
                     raise WorkflowUpdateRPCTimeoutOrCancelledError() from err
                 else:
+                    multiop_failure = (
+                        temporalio.api.errordetails.v1.MultiOperationExecutionFailure()
+                    )
+                    if err.grpc_status.details[0].Unpack(multiop_failure):
+                        status = next(
+                            (
+                                st
+                                for st in multiop_failure.statuses
+                                if (
+                                    st.details
+                                    and not st.details[0].Is(
+                                        temporalio.api.failure.v1.MultiOperationExecutionAborted.DESCRIPTOR
+                                    )
+                                )
+                            ),
+                            None,
+                        )
+                        if status and status.code in RPCStatusCode:
```

Comment:
Should add special handling for `WorkflowAlreadyStartedError` here I think

Reply:
Thank you, that's done now, and in general the error handling and poll loop has been rewritten.